precompiles/utils/src/data.rs

```diff
@@ -368,12 +403,82 @@ impl<T: EvmData> EvmData for Vec<T> {
 			inner_writer = inner_writer.write(inner);
 		}
 
-		let array = Array {
+		let array = OffsetDatum {
+			offset_position,
+			data: inner_writer.data,
+			inner_offset_data: inner_writer.offset_data,
+		};
+
+		writer.offset_data.push(array);
+	}
+}
+
+impl EvmData for Bytes {
+	fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+		let offset_reference = reader.cursor;
+		let array_offset: usize = reader
+			.read::<U256>()
+			.map_err(|_| error("tried to parse array offset out of bounds"))?
+			.try_into()
+			.map_err(|_| error("array offset is too large"))?;
+
+		// We temporarily move the cursor to the offset, we'll set it back afterward.
+		let original_cursor = reader.cursor;
+		reader.cursor = offset_reference + array_offset;
+
+		// Read bytes/string size.
+		let array_size: usize = reader
+			.read::<U256>()
+			.map_err(|_| error("tried to parse bytes/string length out of bounds"))?
+			.try_into()
+			.map_err(|_| error("bytes/string length is too large"))?;
+
+		let range = reader.move_cursor(array_size)?;
+
+		let data = reader
+			.input
+			.get(range)
+			.ok_or_else(|| error("tried to parse bytes/string out of bounds"))?;
+
+		let bytes = Self(data.to_owned());
+
+		// We set back the cursor to its original location.
+		reader.cursor = original_cursor;
+
+		Ok(bytes)
+	}
+
+	fn write(writer: &mut EvmDataWriter, value: Self) {
+		let offset_position = writer.data.len();
+		H256::write(writer, H256::repeat_byte(0xff));
```

Comment:
It may impact prod perf

Reply:
You need to write something anyway to write other (static) stuff after it, and them only after that you can go back and override with the good offset (which is unknown before). The placeholder value is unimportant, and I chosed to use `0xff` such as commenting lines that override the offset allow to clearly see where those should be, which was useful when debuging this.

Reply:
It would be more optimized to write nothing and just advance in a non-initialized memory, like what tokio does internally, but it requires unsafe and it is complex for a gain maybe too low in our case.